File: Contest/TheEmployeeThatWorkedontheLongestTask.java
// https://leetcode.com/problems/the-employee-that-worked-on-the-longest-task/
// @author: anuj0503
class Solution {
    class Person {
        int time;
        int id;

        public Person(int a, int b){
            id = a;
            time = b;
        }
    }
    public int hardestWorker(int n, int[][] logs) {
        PriorityQueue<Person> pq = new PriorityQueue<>(new Comparator<Person>() {
            @Override
            public int compare(Person o1, Person o2) {
                if(o1.time == o2.time) return o1.id - o2.id;
                
                return o2.time - o1.time;
            }
        });

        pq.offer(new Person(logs[0][0], logs[0][1]));

        for(int i = 1; i < logs.length; i++){
            pq.offer(new Person(logs[i][0], logs[i][1] - logs[i - 1][1]));
        }
        
        return pq.poll().id;
    }
}

File: Contest/Weekly Contest 314/TheEmployeeThatWorkedOnTheLongestTask.py
# https://leetcode.com/problems/the-employee-that-worked-on-the-longest-task
# author : @Jiganesh

from typing import List
class Solution:
    
    # TC: O(N)
    # SC: O(1)
    def hardestWorker(self, n: int, logs: List[List[int]]) -> int:
        start = 0
        max_task_time = float("-inf")
        hard_working_employee_id = 0 

        for log in logs:
            
            emp_id , end = log
            task_time = (end-start)

            if task_time > max_task_time:
                max_task_time = task_time
                hard_working_employee_id = emp_id
            
            elif task_time == max_task_time:
                hard_working_employee_id = min(hard_working_employee_id, emp_id)
            
            start = end

        return hard_working_employee_id
